src/keychain/memory.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{Height, Keychain};
+
+use fuel_crypto::borrown::Borrown;
+use fuel_crypto::{Hasher, Keystore, SecretKey, Signer};
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+use core::convert::Infallible;
+use core::ops::{Range, RangeBounds};
+use std::collections::HashMap;
+
+/// Default in-memory implementation of a keychain
+#[derive(Debug, Default, Clone)]
+pub struct MemoryKeychain {
+    keys: HashMap<Range<Height>, SecretKey>,
+}
+
+impl MemoryKeychain {
+    /// Add a new password generated secret to the keychain
+    pub fn insert<H, P>(&mut self, _height: H, password: P)
+    where
+        H: RangeBounds<Height>,
+        P: AsRef<[u8]>,
+    {
+        let seed = Hasher::hash(password);
+        let rng = &mut StdRng::from_seed(*seed);
+        let secret = SecretKey::random(rng);
+
+        // TODO implement range split?
+        self.keys.insert(
+            Range {
```

Comment:
Unless we make the key of the hashmap generic over RangeBounds, this doesn't seem very likely to work. The start and end bounds from RangeBounds can vary between inclusive or exclusive, which would map to different concrete Range types such as `RangeTo` vs `RangeInclusive` vs `Range`.

Reply:
They are always inclusive on the lower bound, and exclusive in the upper bound: [Range](https://doc.rust-lang.org/std/ops/struct.Range.html)

But this will be a TODO even after this PR. It must be implemented with the key rotation: https://github.com/FuelLabs/fuel-bft/issues/10